cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/LogParserTransform.java

```diff
@@ -0,0 +1,183 @@
+  private static final Pattern CLF_LOG_PATTERN = Pattern.compile(
+    //   IP                    id    user      date          request     code     size    referrer    user agent
+    "^([\\d.]+|[:][:][\\d]) (\\S+) (\\S+) \\[([^\\]]+)\\] \"([^\"]+)\" (\\d{3}) ([-\"\\d]+) \"([^\"]+)\" \"([^\"]+)\"");
+  private static final Pattern S3_LOG_PATTERN = Pattern.compile(
+    // bucket owner name   time           ip                    req   reqID operation  key     URI
+    "^(\\S+) (\\S+) \\[(\\p{Print}+)\\] ([\\d.]+|[:][:][\\d]) (\\S+) (\\S+) (\\S+) (\\S+) \"([^\"]+)\" " +
+      //  HTTP stat error code  bytes sent     obj size  time    turn time  referrer         user agent
+      "(\\d{3}) (\\p{Print}+) ([-\"\\d]+) ([-\"\\d]+) ([\\d]+) ([-\"\\d]+) \"(\\p{Print}+)\" \"([^\"]+)\" " +
+      //  version id
+      "(\\p{Print}+)");
+  private static final Pattern REQUEST_PAGE_PATTERN = Pattern.compile("(\\S+)\\s(\\S+).*");
+  private static final int[] S3_INDICES = {9, 3, 4, 17};
+  private static final int[] CLF_INDICES = {5, 4, 1, 9};
+  private static final int S3_REGEX_LENGTH = 18;
+  private static final int CLF_REGEX_LENGTH = 9;
+  private final LogParserConfig config;
+
+  public LogParserTransform(LogParserConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    String log;
+    if ("string".equals(config.inputType)) {
+      log = input.get(config.inputName);
+    } else if ("bytes".equals(config.inputType)) {
+      log = Bytes.toString((ByteBuffer) input.get(config.inputName));
+    } else {
+      LOG.error("Unsupported inputType");
+      return;
+    }
+
+    if (log == null) {
+        LOG.error("Couldn't read schema");
+        return;
+    }
+
+    if ("S3".equals(config.logFormat)) {
+      Matcher logMatcher = S3_LOG_PATTERN.matcher(log);
+      if (!logMatcher.matches() || logMatcher.groupCount() < S3_REGEX_LENGTH) {
+        LOG.error("Couldn't parse log " + log);
+        return;
+      }
+      parseRequest(emitter, logMatcher, S3_INDICES);
+    } else if ("CLF".equals(config.logFormat)) {
+      Matcher logMatcher = CLF_LOG_PATTERN.matcher(log);
+      if (!logMatcher.matches() || logMatcher.groupCount() < CLF_REGEX_LENGTH) {
+        LOG.error("Couldn't parse log " + log);
+        return;
+      }
+      parseRequest(emitter, logMatcher, CLF_INDICES);
+    } else {
+      LOG.error("Log format not currently supported.");
+      return;
+    }
+  }
+
+  /**
+   * Parses a request for the URI, IP, Browser, Device, and Time
+   * @param emitter the emitter to emit parsed information to
+   * @param logMatcher the regex matcher to use
+   * @param indices array of indices that define what position in the regex the fields are,
+   *                 in the order of Request, Time, IP, and User Agent
+   */
+  private void parseRequest(Emitter<StructuredRecord> emitter, Matcher logMatcher, int[] indices) {
```

Comment:
I think it would be better for this method to return ``StructuredRecord``. ``emitter.emit()`` seems better suited to the ``transform`` method. Otherwise, this method's contract is not just *parse*, but also *emit*